### PR TITLE
Add Linux CPU CI for ORT minimal build

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: Add Linux CPU CI for ORT minimal build

**Motivation and Context**
- We need a CI to confirm a PR will not break ORT minimal build
- Also we need run a set of UT under ORT minimal build and test include_ops for the model used in the UT
- The CI runs on Linux (Windows build will not fail if a throw was added to code and build with "--disable_exceptions" build option)

Success CI run,
https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=215464

Failed CI run, (by adding a throw std::runtime_error purposely in the code in this commit https://github.com/microsoft/onnxruntime/commit/a6b2f022ed416fb388bc5e478a49dadf37eec644)
https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=215463

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+          --volume $HOME/.onnx:/home/onnxruntimedev/.onnx \
+          --volume $HOME/.models:/home/onnxruntimedev/.models \
+          -e ALLOW_RELEASED_ONNX_OPSET_ONLY=1 \
+          -e NIGHTLY_BUILD \
+          -e BUILD_BUILDNUMBER \
+          onnxruntimeregistry.azurecr.io/internal/azureml/onnxruntimecpubuild:chn \
+            /opt/python/cp37-cp37m/bin/python3 /onnxruntime_src/tools/ci_build/build.py \
+              --build_dir /build --cmake_generator Ninja \
+              --config Debug MinSizeRel\
+              --skip_submodule_sync \
```
Comment (skottmckay) [inline tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml:38]: Do we need to build 2 different configs for this CI?

File: tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    inputs:
+      containerRegistry: onnxruntimeregistry
+      command: login
+      addPipelineData: false
+  - task: CmdLine@2
+    inputs:
+      script: |
+        mkdir -p $HOME/.models
+        cp $(Build.SourcesDirectory)/onnxruntime/test/testdata/ort_github_issue_4031.onnx $HOME/.models/
+        mkdir -p $HOME/.onnx
```
Comment (skottmckay) [inline tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml:23]: Please document the reasons things like this in comments so it's easier to maintain the build config.

e.g. we copy this from testdata to a separate folder due to the --include_ops_by_model recursively searching the provided folder for all onnx models, and we only want to include ops for this model in the build. 

Would be good to capture somewhere that there's a matching .ort file for this model, and that a unit test validates the build using that model (which explains why we copy this particular model and why it matters).

File: tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+              --config Debug MinSizeRel\
+              --skip_submodule_sync \
+              --build_shared_lib \
+              --parallel \
+              --minimal_build \
+              --disable_ml_ops \
+              --disable_exceptions \
+              --include_ops_by_model /home/onnxruntimedev/.models/ \
+              --cmake_extra_defines PYTHON_INCLUDE_DIR=/opt/python/cp37-cp37m/include/python3.7m PYTHON_LIBRARY=/usr/lib64/librt.so
+      workingDirectory: $(Build.SourcesDirectory)
```
Comment (skottmckay) [inline tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml:46]: Probably don't need this if we're not building a wheel (can't do that as exceptions are disabled)

**Comment by skottmckay** [comment]
What's the overall plan between the Windows CI and the Linux CI for a reduced/minimal build? Should they both test the same things or will they each test different aspects? Ideally we don't have separate 'reduced' and 'minimal' build CIs.

e.g. some potential aspects to test
  - build with just reduced ops (existing Windows CI we could expand)
  - minimal build with reduced ops and exceptions disabled
  - minimal build with reduced ops and python wheel
    - requires exceptions to be enabled but allows testing models using python
  - end-to-end for minimal build with reduced ops
    - using full ORT python package convert a model to ORT format, create minimal build with reduced ops using the optimized model from the conversion, validate the ORT model

File: tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+          --volume $HOME/.models:/home/onnxruntimedev/.models \
+          -e ALLOW_RELEASED_ONNX_OPSET_ONLY=1 \
+          -e NIGHTLY_BUILD \
+          -e BUILD_BUILDNUMBER \
+          onnxruntimeregistry.azurecr.io/internal/azureml/onnxruntimecpubuild:chn \
+            /opt/python/cp37-cp37m/bin/python3 /onnxruntime_src/tools/ci_build/build.py \
+              --build_dir /build --cmake_generator Ninja \
+              --config Debug MinSizeRel\
+              --skip_submodule_sync \
+              --build_shared_lib \
```
[the remark this answers is not in the document]
Comment (gwang-msft) [reply tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml:39]: Yes, it does

---
In reply to: [484226835](https://github.com/microsoft/onnxruntime/pull/5074#discussion_r484226835) [](ancestors = 484226835)

File: tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml (hunk 1 above)
Comment (gwang-msft) [reply tools/ci_build/github/azure-pipelines/linux-cpu-minimal-build-ci-pipeline.yml:38]: Sticks to Debug for now


---
In reply to: [484226560](https://github.com/microsoft/onnxruntime/pull/5074#discussion_r484226560) [](ancestors = 484226560)

**Comment by gwang-msft** [comment]
For now we may have the following
1, The reduced ops CI (windows), we may deprecate this CI in the future and combine what it does with the minimal build CI (the one in the PR)
2. The minimal build with reduced ops and exception disabled (this PR)
3. E2E for minimal build with reduced ops
  - Build full ORT to convert onnx to ort, build a minimal build with reduced opts and python wheel to verify the converted model
  - This will be a temproary CI before we have a onnx_test_runner works with ort format, once we have that, we can have to E2E min build CI as
     Build full ORT to convert onnx to ort, build a minimal build with reduced opts and no exception to verify the converted model


---
In reply to: [688225115](https://github.com/microsoft/onnxruntime/pull/5074#issuecomment-688225115) [](ancestors = 688225115)

**Review by snnn — APPROVED** [review]

**Review by skottmckay — APPROVED** [review]

**Pull request merged**